lib/image/sharp.js

```diff
@@ -0,0 +1,176 @@
+     * will never exceed the original size, and will otherwise be scaled to
+     * fill as much of the requested size as possible without changing the aspect
+     * ratio. Files are generated in the temp folder with a filename made up of the
+     * name property of the size, a '.', and the extension property of the
+     * context object.
+     *
+     * tempFolder: folder where the scaled versions should be written
+     * (required)
+     *
+     * crop: optional object with top, left, width and height properties
+     *
+     * scaledJpegQuality: quality setting for JPEGs (optional; otherwise
+     * you get whatever default was compiled into sharp)
+     *
+     * copyOriginal: if true, copy the "original" image to the tempFolder too,
+     * but do auto-orient it so that iPhone photos etc. work on the web
+     *
+     * All images, including the "original" if copyOriginal is set, are
+     * auto-rotated to the orientation expected by web browsers.
+     *
+     * OUTPUT
+     *
+     * After the operation is complete, the following property of the
+     * context object will be set if the copyOriginal option was set:
+     *
+     * adjustedOriginal: will contain the local filesystem path where the
+     * original was copied (and rotated, if needed).
+     *
+     * @param  {[type]}   context  [description]
+     * @param  {Function} callback [description]
+     * @return {[type]}            [description]
+     */
+
+    convert: function (context, callback) {
+      // This is for a non-animated image
+      const _info = context.info;
+      const isAnimated = _info.animation;
+      const noCrop = {
+        left: 0,
+        top: 0,
+        width: _info.width,
+        height: _info.height
+      };
+      const crop = context.crop ? context.crop : noCrop;
+
+      const pipeline = Sharp(context.workingPath, { animated: isAnimated })
+        .rotate()
+        .extract({
+          left: crop.left,
+          top: crop.top,
+          width: crop.width,
+          height: crop.height
+        });
+
+      const promises = [];
+
+      if (context.copyOriginal) {
+        const copyPath = `${context.tempFolder}/original.${context.extension}`;
+        context.adjustedOriginal = copyPath;
+        promises.push(pipeline.clone().withMetadata().toFile(copyPath));
+      }
+
+      promises.push(sizeOperation());
+
+      Promise.all(promises)
+        .then((res) => {
+          return callback(null);
+        })
+        .catch((err) => {
+          console.error(err);
+          return callback(err);
+        });
+
+      async function sizeOperation() {
+        await Promise.all(
+          context.sizes.map(async (size) => {
+            const sizePath = `${context.tempFolder}/${size.name}.${context.extension}`;
+            const width = Math.min(size.width, context.info.width);
+            const height = Math.min(size.height, context.info.height);
+            if (context.extension === 'jpg') {
```

Comment:
You can reduce code duplication here by holding the pipeline in an intermediate variable so your `if` statement just calls `pipeline.jpeg({ quality })` and doesn't have to duplicate the ret of the logic.

Reply:
Refactored - thanks for the help.